### PR TITLE
feat(codebuild): add functionality to allow using private registry an…

### LINKED ISSUE #2175: CodeBuild project environment should support ImagePullCredentialsType
https://docs.aws.amazon.com/codebuild/latest/APIReference/API_ProjectEnvironment.html#CodeBuild-Type-ProjectEnvironment-imagePullCredentialsType

"When you use a cross-account or private registry image, you must use SERVICE_ROLE credentials. When you use an AWS CodeBuild curated image, you must use CODEBUILD credentials."

```ts
        let buildProject = new codebuild.Project(this, 'build-project', {
            environment: {
                computeType: ComputeType.Large,
                // https://github.com/awslabs/aws-cdk/issues/1325
                buildImage: WindowsBuildImage.fromDockerHub(this.ecrImageUri.stringValue)
            },
            source: new codebuild.CodePipelineSource(),
            buildSpec: 'buildspec.yml'
        });

        (buildProject.node.findChild("Resource") as codebuild.CfnProject)
            .addPropertyOverride("Environment.ImagePullCredentialsType", "SERVICE_ROLE");
```

### PR DESCRIPTION
**Pull request by Kaixiang-AWS**
…d cross-account ECR repository as build image

Fixes #2175

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 1]
```diff
@@ -1287,13 +1314,6 @@ function extendBuildSpec(buildSpec: any, extend: any) {
...
-  return new iam.PolicyStatement()
-    .describe('CodeBuild')
-    .addServicePrincipal('codebuild.amazonaws.com')
-    .addActions(
-      'ecr:GetDownloadUrlForLayer',
-      'ecr:BatchGetImage',
-      'ecr:BatchCheckLayerAvailability'
-    );
+function isECRImage(imageUri: string) {
+  return /^(.+).dkr.ecr.(.+).amazonaws.com[.]{0,1}[a-z]{0,3}\/([^:]+):?.*$/.test(imageUri);
```
Comment (eladb) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: You should ensure imageUri can be parsed using Token.unresolved

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 2]
```diff
@@ -1023,7 +1060,7 @@ export interface IBuildImage {
  *
  * You can also specify a custom image using one of the static methods:
  *
- * - LinuxBuildImage.fromDockerHub(image)
+ * - LinuxBuildImage.fromDockerRegistry(image[, secretsManagerCredential])
```
Comment (eladb) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: @skinny85 @SoManyHs @rix0rrr do you think it makes sense to merge this API with the one we have for ECS?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: In that case, we'd probably need some generic modeling of Docker images, using `@aws-cdk/docker` or something.

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: Discussed offline. We decided not create a general Docker image API. Given that, I believe this code is fine - what do you think @eladb ?

**Comment by skinny85** [comment]
Hey @Kaixiang-AWS ,

apologies for not helping you with this PR sooner. I promise I'll improve from now on 🤞. I see there are some conflicts with master, and I believe some accidental changes in the `docker-assets` package that should not be there currently in the diff. Would you mind merging/rebasing (whichever you prefer) with the newest tip of the master branch, and we'll pick it up from there?

Thanks,
Adam

**Comment by Kaixiang-AWS** [comment]
> Hey @Kaixiang-AWS ,
> 
> apologies for not helping you with this PR sooner. I promise I'll improve from now on 🤞. I see there are some conflicts with master, and I believe some accidental changes in the `docker-assets` package that should not be there currently in the diff. Would you mind merging/rebasing (whichever you prefer) with the newest tip of the master branch, and we'll pick it up from there?
> 
> Thanks,
> Adam

I'm working on the rebasing and will have a PR out soon

**Comment by skinny85** [comment]
@Kaixiang-AWS ,

can we close this PR, now that #3049 has been opened?

**Comment by Kaixiang-AWS** [comment]
tracking in #3049

**Pull request closed without merging**